devices/HumanDataCollector/HumanDataCollector.cpp

```diff
@@ -0,0 +1,663 @@
+                yError() << LogPrefix << "Failed to open port " << jointTorquesDataPortName;
+                askToStop();
+            }
+
+        }
+
+
+    }
+
+
+    // =========================================================
+    // Get data from different interfaces from attached devices
+    // =========================================================
+
+    // Get data from IHumanState interface of HumanStateProvider
+    if (pImpl->isStateProviderDevice) {
+
+        // Base Quantities
+        pImpl->baseName = pImpl->iHumanState->getBaseName();
+        pImpl->basePosition = pImpl->iHumanState->getBasePosition();
+        pImpl->baseOrientation = pImpl->iHumanState->getBaseOrientation();
+        pImpl->baseVelocity = pImpl->iHumanState->getBaseVelocity();
+
+        // Joint Quantities
+        pImpl->stateNumberOfJoints = pImpl->iHumanState->getNumberOfJoints();
+        pImpl->stateJointNames = pImpl->iHumanState->getJointNames();
+        pImpl->jointPositions = pImpl->iHumanState->getJointPositions();
+        pImpl->jointVelocities = pImpl->iHumanState->getJointVelocities();
+
+        // CoM Quantities
+        pImpl->comPosition = pImpl->iHumanState->getCoMPosition();
+        pImpl->comVelocity = pImpl->iHumanState->getCoMVelocity();
+        pImpl->comProperAccInBaseFrame = pImpl->iHumanState->getCoMProperAccelerationExpressedInBaseFrame();
+        pImpl->comProperAccInWorldFrame = pImpl->iHumanState->getCoMProperAccelerationExpressedInWorldFrame();
+
+    }
+
+    // Get data from IHumanWrench interface of HumanWrenchProvider
+    if (pImpl->isWrenchProviderDevice) {
+
+        pImpl->numberOfWrenchMeasurementSources = pImpl->iHumanWrenchMeasurements->getNumberOfWrenchSources();
+        pImpl->wrenchMeasurementSourceNames = pImpl->iHumanWrenchMeasurements->getWrenchSourceNames();
+        pImpl->wrenchMeasurementValues = pImpl->iHumanWrenchMeasurements->getWrenches();
+
+    }    
+
+    if (pImpl->isDynamicsEstimatorDevice) {
+
+        // Get data from IHumanWrench interface of HumanDynamicsEstimator
+        // NOTE: The wrench values coming from HumanDynamicsEstimators are (offsetRemovedWrenchMeasurements & WrenchEstimates) of each link
+        pImpl->numberOfWrenchEstimateSources = pImpl->iHumanWrenchEstimates->getNumberOfWrenchSources();
+        pImpl->wrenchEstimateSourceNames = pImpl->iHumanWrenchEstimates->getWrenchSourceNames();
+        pImpl->wrenchEstimateValues = pImpl->iHumanWrenchEstimates->getWrenches();
+
+        // Get data from IHumanDynamics interface of HumanDynamicsEstimator
+        pImpl->dynamicsNumberOfJoints = pImpl->iHumanDynamics->getNumberOfJoints();
+        pImpl->dynamicsJointNames = pImpl->iHumanDynamics->getJointNames();
+        pImpl->jointTorques = pImpl->iHumanDynamics->getJointTorques();
+
+    }
+
+    // Flip firstData flag false
+    if (pImpl->firstData) {
+        pImpl->firstData = false;
+    }
```

Comment:
Maybe this could be moved above?

```cpp
if (pImpl->firstData) {
    pImpl->firstData = false;
    // ...
}
```

Reply:
@diegoferigo I did not get this

Reply:
Talked f2f. Move it in the top right after the execution enters the first if.